ckanext/etsin/mappers/datacite.py

```diff
@@ -0,0 +1,257 @@
+'''
+Map Datacite 3.1 and 4.0 dicts to Metax values
+'''
+
+from lxml import objectify
+
+from ..utils import validate_6391, get_language_identifier, is_uri
+
+import logging
+log = logging.getLogger(__name__)
+
+# Mapper receives an LXML element and maps it to Metax format
+
+
+def datacite_mapper(xml):
+
+    from lxml import etree
+
+    # Strip element prefixes
+    xml = _strip_prefixes(xml)
+
+    # Start with an empty slate
+    package_dict = {}
+
+    # Map identifier
+    # According to DataCite 4.0 schema, IdentifierType should always be "DOI",
+    # but OpenAire seems to use URLs as well
+    identifier = xml.find('.//identifier')
+    identifier_type = xml.find('.//identifier').get('identifierType')
+    if identifier_type == "URL":
+        package_dict['preferred_identifier'] = identifier.text
+    elif identifier_type == "DOI":
+        package_dict['preferred_identifier'] = "https://dx.doi.org/" + \
+            identifier.text
+    else:
+        package_dict['preferred_identifier'] = ""
+
+    # Map creators
+    package_dict['creator'] = []
+    for creator in xml.findall('.//creator'):
+        person = _get_person(creator)
+        if not person:
+            continue
+        package_dict['creator'].append(person)
+
+    # Map language
+    language = xml.find('.//identifier')
+    if language is not None:
+        # Language should be either ISO 639-1 or IETF BCP 47. The first two
+        # characters of IETF BCP 47 should be same as ISO 639-1 code.
+        if validate_6391(language[0:2]):
+            language = get_language_identifier(language[0:2])
+            package_dict['language'] = language
+        else:
+            language = "default"
+    else:
+        # Use "default" for title and description language codes, but don't
+        # save it in language field
+        language = "default"
+
+    # Map title
+    # In case of multiple primary titles, pick only the first one
+    for title in xml.findall('.//title'):
+        title_type = xml.find('.//title').get('titleType')
+        if not title_type:
+            # Title is primary when there's no titleType
+            package_dict['title'] = [{language: title.text}]
```

Comment:
title's cardinality is 1, so array should be left out?

Reply:
Right.